In [16]:
# Installing necessary Packages

!pip install pyarrow # this helps to convert file to parque files
!pip install azure-storage-blob # this is a function of azure that will allow us  communicate with azure


Defaulting to user installation because normal site-packages is not writeable


ERROR: Invalid requirement: '#'

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


Defaulting to user installation because normal site-packages is not writeable


ERROR: Invalid requirement: '#'

[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
!pip install python-dotenv

Defaulting to user installation because normal site-packages is not writeable



[notice] A new release of pip is available: 24.0 -> 24.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [18]:
### Import necessary libraries

import pandas as pd


In [19]:
import os
import io
import dotenv


In [20]:
from dotenv import load_dotenv
from azure.storage.blob import BlobServiceClient, BlobClient

#### Extraction

In [21]:


ziko_df = pd.read_csv(r"ziko_logistics_data.csv")

In [22]:
ziko_df.head()

Transaction_ID                           Date  Customer_ID  Product_ID  \
0             200  2020-01-01 20:32:25.945945945         1086         536   
1             321  2020-01-02 06:55:08.108108108         1078         523   
2             989  2020-01-06 08:12:58.378378378         1077         535   
3             682  2020-01-07 22:03:14.594594594         1027         546   
4             484  2020-01-07 07:08:06.486486486         1052         556   

   Quantity  Unit_Price   Total_Cost  Discount_Rate Sales_Channel  \
0         3  120.436821  8265.374549           0.20        Online   
1         6  475.724994  4047.850479            NaN      Reseller   
2         3  146.400556          NaN           0.05        Direct   
3         6   19.373194  8194.281993            NaN      Reseller   
4         8  193.221313  8331.329249           0.20        Direct   

  Order_Priority  ... Return_Reason   Payment_Type Taxable Region  Country  \
0           High  ...    Wrong Item  Wire Transfer   False   West   Canada   
1       Critical  ...       Damaged         PayPal    True  South   Mexico   
2       Critical  ...       Damaged         PayPal    True   West   Canada   
3         Medium  ...    Wrong Item           Cash    True  South   Canada   
4            Low  ...          Late           Cash   False  South   Mexico   

  Customer_Name   Customer_Phone               Customer_Email  \
0  Customer 200  +1-652-572-9306  customer.200.78@example.com   
1  Customer 321  +1-311-186-5760   customer.321.13@sample.com   
2  Customer 989  +1-922-606-9032  customer.989.99@example.com   
3  Customer 682  +1-237-853-5808     customer.682.66@demo.com   
4  Customer 484  +1-986-360-9109    customer.484.3@sample.com   

               Customer_Address Product_List_Title  
0   275 Second St, Phoenix, USA         Product 53  
1   478 Third St, New York, USA         Product 33  
2   843 Second St, Phoenix, USA          Product 6  
3     153 Main St, Phoenix, USA         Product 68  
4  264 Second St, New York, USA         Product 89  

[5 rows x 25 columns]

In [23]:
ziko_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Transaction_ID         1005 non-null   int64  
 1   Date                   1005 non-null   object 
 2   Customer_ID            1005 non-null   int64  
 3   Product_ID             1005 non-null   int64  
 4   Quantity               1005 non-null   int64  
 5   Unit_Price             904 non-null    float64
 6   Total_Cost             905 non-null    float64
 7   Discount_Rate          714 non-null    float64
 8   Sales_Channel          1005 non-null   object 
 9   Order_Priority         1005 non-null   object 
 10  Warehouse_Code         1005 non-null   object 
 11  Ship_Mode              1005 non-null   object 
 12  Delivery_Status        1005 non-null   object 
 13  Customer_Satisfaction  1005 non-null   object 
 14  Item_Returned          1005 non-null   bool   
 15  Retu

In [24]:
print (ziko_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Transaction_ID         1005 non-null   int64  
 1   Date                   1005 non-null   object 
 2   Customer_ID            1005 non-null   int64  
 3   Product_ID             1005 non-null   int64  
 4   Quantity               1005 non-null   int64  
 5   Unit_Price             904 non-null    float64
 6   Total_Cost             905 non-null    float64
 7   Discount_Rate          714 non-null    float64
 8   Sales_Channel          1005 non-null   object 
 9   Order_Priority         1005 non-null   object 
 10  Warehouse_Code         1005 non-null   object 
 11  Ship_Mode              1005 non-null   object 
 12  Delivery_Status        1005 non-null   object 
 13  Customer_Satisfaction  1005 non-null   object 
 14  Item_Returned          1005 non-null   bool   
 15  Retu

In [25]:
ziko_df.columns

Index(['Transaction_ID', 'Date', 'Customer_ID', 'Product_ID', 'Quantity',
       'Unit_Price', 'Total_Cost', 'Discount_Rate', 'Sales_Channel',
       'Order_Priority', 'Warehouse_Code', 'Ship_Mode', 'Delivery_Status',
       'Customer_Satisfaction', 'Item_Returned', 'Return_Reason',
       'Payment_Type', 'Taxable', 'Region', 'Country', 'Customer_Name',
       'Customer_Phone', 'Customer_Email', 'Customer_Address',
       'Product_List_Title'],
      dtype='object')

#### Exploration

In [26]:
ziko_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Transaction_ID         1005 non-null   int64  
 1   Date                   1005 non-null   object 
 2   Customer_ID            1005 non-null   int64  
 3   Product_ID             1005 non-null   int64  
 4   Quantity               1005 non-null   int64  
 5   Unit_Price             904 non-null    float64
 6   Total_Cost             905 non-null    float64
 7   Discount_Rate          714 non-null    float64
 8   Sales_Channel          1005 non-null   object 
 9   Order_Priority         1005 non-null   object 
 10  Warehouse_Code         1005 non-null   object 
 11  Ship_Mode              1005 non-null   object 
 12  Delivery_Status        1005 non-null   object 
 13  Customer_Satisfaction  1005 non-null   object 
 14  Item_Returned          1005 non-null   bool   
 15  Retu

In [27]:
ziko_df.head()

Transaction_ID                           Date  Customer_ID  Product_ID  \
0             200  2020-01-01 20:32:25.945945945         1086         536   
1             321  2020-01-02 06:55:08.108108108         1078         523   
2             989  2020-01-06 08:12:58.378378378         1077         535   
3             682  2020-01-07 22:03:14.594594594         1027         546   
4             484  2020-01-07 07:08:06.486486486         1052         556   

   Quantity  Unit_Price   Total_Cost  Discount_Rate Sales_Channel  \
0         3  120.436821  8265.374549           0.20        Online   
1         6  475.724994  4047.850479            NaN      Reseller   
2         3  146.400556          NaN           0.05        Direct   
3         6   19.373194  8194.281993            NaN      Reseller   
4         8  193.221313  8331.329249           0.20        Direct   

  Order_Priority  ... Return_Reason   Payment_Type Taxable Region  Country  \
0           High  ...    Wrong Item  Wire Transfer   False   West   Canada   
1       Critical  ...       Damaged         PayPal    True  South   Mexico   
2       Critical  ...       Damaged         PayPal    True   West   Canada   
3         Medium  ...    Wrong Item           Cash    True  South   Canada   
4            Low  ...          Late           Cash   False  South   Mexico   

  Customer_Name   Customer_Phone               Customer_Email  \
0  Customer 200  +1-652-572-9306  customer.200.78@example.com   
1  Customer 321  +1-311-186-5760   customer.321.13@sample.com   
2  Customer 989  +1-922-606-9032  customer.989.99@example.com   
3  Customer 682  +1-237-853-5808     customer.682.66@demo.com   
4  Customer 484  +1-986-360-9109    customer.484.3@sample.com   

               Customer_Address Product_List_Title  
0   275 Second St, Phoenix, USA         Product 53  
1   478 Third St, New York, USA         Product 33  
2   843 Second St, Phoenix, USA          Product 6  
3     153 Main St, Phoenix, USA         Product 68  
4  264 Second St, New York, USA         Product 89  

[5 rows x 25 columns]

In [28]:
### Creating a copy of the dataframe

ziko_df2 = ziko_df.copy()

In [43]:
ziko_df2.info

<bound method DataFrame.info of       Transaction_ID                          Date  Customer_ID  Product_ID  \
0                200 2020-01-01 20:32:25.945945945         1086         536   
1                321 2020-01-02 06:55:08.108108108         1078         523   
2                989 2020-01-06 08:12:58.378378378         1077         535   
3                682 2020-01-07 22:03:14.594594594         1027         546   
4                484 2020-01-07 07:08:06.486486486         1052         556   
...              ...                           ...          ...         ...   
1000             250 2020-01-05 02:22:42.162162162         1095         533   
1001             412 2020-01-07 11:27:34.054054054         1008         542   
1002             816 2020-01-06 02:48:38.918918918         1060         511   
1003             702 2020-01-09 08:38:55.135135135         1054         504   
1004               4 2020-01-03 10:22:42.162162162         1018         545   

      Quantity  Uni

#### Issues Identified

1. Missing values:

    Unit_Price             904 non-null    float64
    Total_Cost             905 non-null    float64
    Discount_Rate          714 non-null    float64
    Return_Reason          904 non-null    object 

2. Wrong data type:

    Date                   1005 non-null   object

#### Transformation

In [57]:
ziko_df2.fillna ({
    "Unit_Price" : ziko_df2["Unit_Price"].mean(),
    "Total_Cost" : ziko_df2["Total_Cost"].mean(),
    "Discount_Rate" : 0.0,
    "Return_Reason" : "Unknown"
}, inplace=True)

In [58]:
ziko_df2.head()

Transaction_ID                          Date  Customer_ID  Product_ID  \
0             200 2020-01-01 20:32:25.945945945         1086         536   
1             321 2020-01-02 06:55:08.108108108         1078         523   
2             989 2020-01-06 08:12:58.378378378         1077         535   
3             682 2020-01-07 22:03:14.594594594         1027         546   
4             484 2020-01-07 07:08:06.486486486         1052         556   

   Quantity  Unit_Price   Total_Cost  Discount_Rate Sales_Channel  \
0         3  120.436821  8265.374549           0.20        Online   
1         6  475.724994  4047.850479           0.00      Reseller   
2         3  146.400556  5096.553818           0.05        Direct   
3         6   19.373194  8194.281993           0.00      Reseller   
4         8  193.221313  8331.329249           0.20        Direct   

  Order_Priority  ... Return_Reason   Payment_Type Taxable Region  Country  \
0           High  ...    Wrong Item  Wire Transfer   False   West   Canada   
1       Critical  ...       Damaged         PayPal    True  South   Mexico   
2       Critical  ...       Damaged         PayPal    True   West   Canada   
3         Medium  ...    Wrong Item           Cash    True  South   Canada   
4            Low  ...          Late           Cash   False  South   Mexico   

  Customer_Name   Customer_Phone               Customer_Email  \
0  Customer 200  +1-652-572-9306  customer.200.78@example.com   
1  Customer 321  +1-311-186-5760   customer.321.13@sample.com   
2  Customer 989  +1-922-606-9032  customer.989.99@example.com   
3  Customer 682  +1-237-853-5808     customer.682.66@demo.com   
4  Customer 484  +1-986-360-9109    customer.484.3@sample.com   

               Customer_Address Product_List_Title  
0   275 Second St, Phoenix, USA         Product 53  
1   478 Third St, New York, USA         Product 33  
2   843 Second St, Phoenix, USA          Product 6  
3     153 Main St, Phoenix, USA         Product 68  
4  264 Second St, New York, USA         Product 89  

[5 rows x 25 columns]

In [70]:
### Converting Date data type to datetime

ziko_df2["Date"] = pd.to_datetime(ziko_df2["Date"])

In [71]:
ziko_df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 25 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Transaction_ID         1005 non-null   int64         
 1   Date                   1005 non-null   datetime64[ns]
 2   Customer_ID            1005 non-null   int64         
 3   Product_ID             1005 non-null   int64         
 4   Quantity               1005 non-null   int64         
 5   Unit_Price             1005 non-null   float64       
 6   Total_Cost             1005 non-null   float64       
 7   Discount_Rate          1005 non-null   float64       
 8   Sales_Channel          1005 non-null   object        
 9   Order_Priority         1005 non-null   object        
 10  Warehouse_Code         1005 non-null   object        
 11  Ship_Mode              1005 non-null   object        
 12  Delivery_Status        1005 non-null   object        
 13  Cus

In [72]:
### Creating Tables

# Customer Table

customer = ziko_df2[[ "Customer_ID", "Customer_Name", "Customer_Phone", "Customer_Email", "Customer_Address"]].copy().drop_duplicates().reset_index(drop=True)

customer.head()

Customer_ID Customer_Name   Customer_Phone               Customer_Email  \
0         1086  Customer 200  +1-652-572-9306  customer.200.78@example.com   
1         1078  Customer 321  +1-311-186-5760   customer.321.13@sample.com   
2         1077  Customer 989  +1-922-606-9032  customer.989.99@example.com   
3         1027  Customer 682  +1-237-853-5808     customer.682.66@demo.com   
4         1052  Customer 484  +1-986-360-9109    customer.484.3@sample.com   

               Customer_Address  
0   275 Second St, Phoenix, USA  
1   478 Third St, New York, USA  
2   843 Second St, Phoenix, USA  
3     153 Main St, Phoenix, USA  
4  264 Second St, New York, USA

In [73]:

customer.shape

(1000, 5)

In [74]:
### Product Table

product = ziko_df2[[ "Product_ID", "Product_List_Title","Quantity", "Unit_Price", "Discount_Rate"]].copy().drop_duplicates().reset_index(drop=True)

product.head()

Product_ID Product_List_Title  Quantity  Unit_Price  Discount_Rate
0         536         Product 53         3  120.436821           0.20
1         523         Product 33         6  475.724994           0.00
2         535          Product 6         3  146.400556           0.05
3         546         Product 68         6   19.373194           0.00
4         556         Product 89         8  193.221313           0.20

In [75]:
# Transaction_fact_table

transaction_fact = ziko_df2.merge( customer, on = ["Customer_ID", "Customer_Name", "Customer_Phone", "Customer_Email", "Customer_Address"], how = "left" ) \
                            .merge( product, on = ["Product_ID", "Product_List_Title","Quantity", "Unit_Price", "Discount_Rate"], how = "left" ) \
                            [['Transaction_ID', 'Date', 'Customer_ID', 'Product_ID', 'Total_Cost', 'Sales_Channel','Order_Priority', 'Warehouse_Code', 'Ship_Mode', \
                            'Delivery_Status', 'Customer_Satisfaction', 'Item_Returned', 'Return_Reason','Payment_Type', 'Taxable', 'Region', 'Country' ]]

In [ ]:
# Transaction_fact_table

transaction_fact = ziko_df2.merge( customer, on = ["Customer_ID", "Customer_Name", "Customer_Phone", "Customer_Email", "Customer_Address"], how = "left" ) \
                            .merge( product, on = ["Product_ID", "Product_List_Title","Quantity", "Unit_Price", "Discount_Rate"], how = "left" ) \
                            [['Transaction_ID', 'Date', 'Customer_ID', 'Product_ID', 'Total_Cost', 'Sales_Channel','Order_Priority', 'Warehouse_Code', 'Ship_Mode', \
                            'Delivery_Status', 'Customer_Satisfaction', 'Item_Returned', 'Return_Reason','Payment_Type', 'Taxable', 'Region', 'Country' ]]

In [76]:
transaction_fact.head()

Transaction_ID                          Date  Customer_ID  Product_ID  \
0             200 2020-01-01 20:32:25.945945945         1086         536   
1             321 2020-01-02 06:55:08.108108108         1078         523   
2             989 2020-01-06 08:12:58.378378378         1077         535   
3             682 2020-01-07 22:03:14.594594594         1027         546   
4             484 2020-01-07 07:08:06.486486486         1052         556   

    Total_Cost Sales_Channel Order_Priority Warehouse_Code  Ship_Mode  \
0  8265.374549        Online           High           WH-3      2-Day   
1  4047.850479      Reseller       Critical           WH-1  Overnight   
2  5096.553818        Direct       Critical           WH-1  Overnight   
3  8194.281993      Reseller         Medium           WH-1    Express   
4  8331.329249        Direct            Low           WH-2      2-Day   

  Delivery_Status Customer_Satisfaction  Item_Returned Return_Reason  \
0       Cancelled               Neutral          False    Wrong Item   
1       Backorder             Satisfied           True       Damaged   
2         Pending           Unsatisfied           True       Damaged   
3         Pending           Unsatisfied          False    Wrong Item   
4       Delivered             Satisfied           True          Late   

    Payment_Type  Taxable Region Country  
0  Wire Transfer    False   West  Canada  
1         PayPal     True  South  Mexico  
2         PayPal     True   West  Canada  
3           Cash     True  South  Canada  
4           Cash    False  South  Mexico

In [77]:
# Temporary Loading

customer.to_csv(r"Dataset/Customer.csv", index=False)
product.to_csv(r"Dataset/Product.csv", index=False)
transaction_fact.to_csv(r"Dataset/Transaction_Fact.csv", index=False)

#### Loading The Data

In [78]:

load_dotenv()

# Setting up the Azure connection

connect_str = os.getenv("AzureCS")
blob_service_client = BlobServiceClient.from_connection_string(connect_str)

# validating the parameters that is needed to connect to the container
container_name = os.getenv("AzureCN")
container_client = blob_service_client.get_container_client(container_name)


In [79]:
# Create a function that would load the data into Azure blob storage as a Parquet file

def upload_df_to_blob_as_parquet(df, container_client, blob_name):
    buffer = io.BytesIO()
    df.to_parquet(buffer, index=False)
    buffer.seek(0)
    blob_client = container_client.get_blob_client(blob_name)
    blob_client.upload_blob(buffer, blob_type = "BlockBlob", overwrite = True)
    print (f"{blob_name} uploaded to Blob Storage successfully")

In [81]:
upload_df_to_blob_as_parquet(customer, container_client, "rawdata/customer.parquet")
upload_df_to_blob_as_parquet(product, container_client, "rawdata/product.parquet")
upload_df_to_blob_as_parquet(transaction_fact, container_client, "rawdata/transaction_fact.parquet")

rawdata/customer.parquet uploaded to Blob Storage successfully
rawdata/product.parquet uploaded to Blob Storage successfully
rawdata/transaction_fact.parquet uploaded to Blob Storage successfully


In [82]:
transaction_fact.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1005 entries, 0 to 1004
Data columns (total 17 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   Transaction_ID         1005 non-null   int64         
 1   Date                   1005 non-null   datetime64[ns]
 2   Customer_ID            1005 non-null   int64         
 3   Product_ID             1005 non-null   int64         
 4   Total_Cost             1005 non-null   float64       
 5   Sales_Channel          1005 non-null   object        
 6   Order_Priority         1005 non-null   object        
 7   Warehouse_Code         1005 non-null   object        
 8   Ship_Mode              1005 non-null   object        
 9   Delivery_Status        1005 non-null   object        
 10  Customer_Satisfaction  1005 non-null   object        
 11  Item_Returned          1005 non-null   bool          
 12  Return_Reason          1005 non-null   object        
 13  Pay

In [ ]:
# transaction_fact["Date"] = transaction_fact["Date"].astype["datetime64(us)"]